File: dyros_robot_controller/examples/python/fr3_controller.py
# ============================ FR3 Manipulator Controller ============================
import os
import numpy as np
from typing import Dict
from pynput import keyboard  # Non-blocking, global keyboard listener

from drc.manipulator.robot_data import RobotData
from drc.manipulator.robot_controller import RobotController


class FR3Controller:
    """
    FR3Controller
    -------------
    Minimal example of using Dyros Robot Controller for a 7-DoF FR3 arm.

    - Keeps internal state (q, qdot, EE pose/velocity).
    - Switches control modes via global hotkeys (1/2/3).
    - On every mode switch, snapshots current state as initial references.

    Parameters
    ----------
    dt : float
        Simulation/control timestep [s].
    """

    def __init__(self, dt: float):
        # --- Core configuration/state ---
        self.dt: float = dt
        self.sim_time: float = 0.0

        # Paths to URDF/SRDF (model files)
        root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        urdf_file_path = os.path.join(root, "robots", "fr3", "fr3.urdf")
        srdf_file_path = os.path.join(root, "robots", "fr3", "fr3.srdf")

        # Instantiate dyros robot model/controller
        self.robot_data = RobotData(urdf_path=urdf_file_path, srdf_path=srdf_file_path)
        self.robot_controller = RobotController(dt=self.dt, robot_data=self.robot_data)

        # Degree of freedom
        self.dof: int = self.robot_data.get_dof()

        # --- Joint space states (measured/desired/snapshots) ---
        self.q            = np.zeros(self.dof)  # measured joints
        self.qdot         = np.zeros(self.dof)  # measured joint velocities
        self.q_desired    = np.zeros(self.dof)  # desired joints
        self.qdot_desired = np.zeros(self.dof)  # desired joint velocities
        self.q_init       = np.zeros(self.dof)  # snapshot at mode entry
        self.qdot_init    = np.zeros(self.dof)  # snapshot at mode entry
        self.tau_desired  = np.zeros(self.dof)  # output torques

        # --- Task space (end-effector) states (measured/desired/snapshots) ---
        self.ee_link_name = "fr3_link8"
        self.x            = np.eye(4)   # measured EE pose (4x4)
        self.xdot         = np.zeros(6) # measured EE twist [vx, vy, vz, wx, wy, wz]
        self.x_desired    = np.eye(4)   # desired EE pose
        self.xdot_desired = np.zeros(6) # desired EE twist
        self.x_init       = np.eye(4)   # snapshot at mode entry
        self.xdot_init    = np.zeros(6) # snapshot at mode entry

        # --- Mode bookkeeping (naming/style unified with XLSController) ---
        self.control_mode: str = "Home"
        self.is_control_mode_changed: bool = True
        self.control_start_time: float = 0.0
        
        # Print FR3 URDF info
        print("info:")
        print(self.robot_data.get_verbose())

        # Global keyboard listener (non-blocking)
        self._listener = keyboard.Listener(on_press=self._on_key_press)
        self._listener.daemon = True
        self._listener.start()
        print("[FR3Controller] Keyboard: [1]=Home, [2]=QPIK, [3]=Gravity Compensation")

    def update_model(self, current_time: float, qpos_dict: Dict[str, float], qvel_dict: Dict[str, float]) -> None:
        """
        Update internal model/state from simulator readings.

        Parameters
        ----------
        current_time : float
            Simulator time [s].
        qpos_dict : Dict[str, float]
            Mapping of joint name -> position (rad).
        qvel_dict : Dict[str, float]
            Mapping of joint name -> velocity (rad/s).
        """
        # Time update (shared convention)
        self.sim_time = current_time

        # Read joint states (joint naming must match the simulator)
        for i in range(self.dof):
            jn = f"fr3_joint{i+1}"
            self.q[i] = qpos_dict[jn]
            self.qdot[i] = qvel_dict[jn]

        # Push to dyros robot model and cache EE pose/twist
        self.robot_data.update_state(self.q, self.qdot)
        self.x = self.robot_data.get_pose(self.ee_link_name)
        self.xdot = self.robot_data.get_velocity(self.ee_link_name)

    def compute(self) -> Dict[str, float]:
        """
        Compute control command for the current mode.

        Returns
        -------
        Dict[str, float]
            Actuator torque map: joint name -> tau [Nm].
        """
        # One-time init per mode entry
        if self.is_control_mode_changed:
            self.is_control_mode_changed = False
            self.control_start_time = self.sim_time

            # Snapshot current measured states as new references
            self.q_init = self.q.copy()
            self.qdot_init = self.qdot.copy()
            self.x_init = self.x.copy()
            self.xdot_init = self.xdot.copy()

            # Reset desired trajectories to snapshots
            self.q_desired = self.q_init.copy()
            self.qdot_desired = np.zeros_like(self.qdot_init)
            self.x_desired = self.x_init.copy()
            self.xdot_desired = np.zeros_like(self.xdot_init)

        # --- Mode: Home (joint-space cubic to a predefined posture) ---
        if self.control_mode == "Home":
            q_home = np.array([0.0, 0.0, 0.0, -np.pi/2.0, 0.0, np.pi/2.0, np.pi/4.0])
            self.q_desired = self.robot_controller.move_joint_position_cubic(
                q_target=q_home,
                qdot_target=np.zeros(self.dof),
                q_init=self.q_init,
                qdot_init=self.qdot_init,
                init_time=self.control_start_time,
                current_time=self.sim_time,
                duration=3.0,
            )
            self.qdot_desired = self.robot_controller.move_joint_velocity_cubic(
                q_target=q_home,
                qdot_target=np.zeros(self.dof),
                q_init=self.q_init,
                qdot_init=self.qdot_init,
                init_time=self.control_start_time,
                current_time=self.sim_time,
                duration=3.0,
            )
            self.tau_desired = self.robot_controller.move_joint_torque_step(
                q_target=self.q_desired, qdot_target=self.qdot_desired
            )

        # --- Mode: QPIK (task-space, QP-based IK with cubic profiling) ---
        elif self.control_mode == "QPIK":
            target_x = self.x_init.copy()
            target_x[:3, 3] = target_x[:3, 3] + np.array([0.0, 0.1, 0.1])  # +10 cm in Y and Z

            self.qdot_desired = self.robot_controller.QPIK_cubic(
                x_target=target_x,
                xdot_target=np.zeros(6),
                x_init=self.x_init,
                xdot_init=self.xdot_init,
                init_time=self.control_start_time,
                current_time=self.sim_time,
                duration=2.0,
                link_name=self.ee_link_name,
            )
            # Simple Euler integrate desired joint positions from qdot_desired
            self.q_desired = self.q + self.qdot_desired * self.dt

            # Map (q, qdot) -> torque (PD + gravity)
            self.tau_desired = self.robot_controller.move_joint_torque_step(
                q_target=self.q_desired, qdot_target=self.qdot_desired
            )

        # --- Mode: Gravity Compensation (no tracking) ---
        elif self.control_mode == "Gravity Compensation":
            self.tau_desired = self.robot_data.get_gravity()

        # Format output for simulator actuators
        return {f"fr3_joint{i+1}": float(self.tau_desired[i]) for i in range(self.dof)}

    def _set_mode(self, control_mode: str) -> None:
        """
        Switch control mode and trigger per-mode re-initialization.

        Parameters
        ----------
        control_mode : str
            One of {"Home", "QPIK", "Gravity Compensation"}.
        """
        self.is_control_mode_changed = True
        self.control_mode = control_mode
        print(f"Control Mode Changed: {self.control_mode}")

    def _on_key_press(self, key) -> None:
        """Global hotkeys for mode switching: 1=Home, 2=QPIK, 3=Gravity Compensation."""
        try:
            if key.char == '1':
                self._set_mode("Home")
            elif key.char == '2':
                self._set_mode("QPIK")
            elif key.char == '3':
                self._set_mode("Gravity Compensation")
        except AttributeError:
            # Ignore non-character keys
            pass

File: dyros_robot_controller/examples/C++/src/fr3_controller.cpp
#include "fr3_controller.hpp"

#ifndef ROBOTS_DIRECTORY
#  error "ROBOTS_DIRECTORY is not defined. Define in CMake: -DROBOTS_DIRECTORY=\"${CMAKE_SOURCE_DIR}/../robots\""
#endif

FR3Controller::FR3Controller(const double dt)
: dt_(dt)
{
  // Paths to URDF/SRDF (model files)
  const std::string urdf = std::string(ROBOTS_DIRECTORY) + "/fr3/" + "fr3.urdf";
  const std::string srdf = std::string(ROBOTS_DIRECTORY) + "/fr3/" + "fr3.srdf";

  // Instantiate dyros robot model/controller
  robot_data_ = std::make_shared<drc::Manipulator::RobotData>(urdf, srdf);
  robot_controller_ = std::make_shared<drc::Manipulator::RobotController>(dt_, robot_data_);

  // Degree of freedom
  dof_ = robot_data_->getDof();

  // --- Joint-space states (initialize to zero/snapshot defaults) ---
  q_.setZero(dof_);      
  qdot_.setZero(dof_);
  q_desired_.setZero(dof_);  
  qdot_desired_.setZero(dof_);
  q_init_.setZero(dof_);
  qdot_init_.setZero(dof_);
  tau_desired_.setZero(dof_);

  // --- Task-space states (EE pose/twist and snapshots) ---
  x_.setIdentity();         
  xdot_.setZero();          
  x_desired_.setIdentity(); 
  xdot_desired_.setZero();    
  x_init_.setIdentity();    
  xdot_init_.setZero();    

  // Print FR3 URDF info
  std::cout << "info: \n" << robot_data_->getVerbose() << std::endl; 

  // Global keyboard listener (non-blocking)
  startKeyListener_();
}

FR3Controller::~FR3Controller() 
{
  stopKeyListener_();
}

void FR3Controller::updateModel(const double current_time,
                                const std::unordered_map<std::string, double>& qpos_dict,
                                const std::unordered_map<std::string, double>& qvel_dict)
{
  // Time update (shared convention)
  sim_time_ = current_time;

  // Read joint states (joint naming must match the simulator)
  for (size_t i = 0; i < dof_; ++i) 
  {
    const std::string key = "fr3_joint" + std::to_string(i+1);
    q_(i)  = qpos_dict.at(key);
    qdot_(i) = qvel_dict.at(key);
  }

  // Push to dyros robot model and cache EE pose/twist
  robot_data_->updateState(q_, qdot_);
  x_  = robot_data_->getPose(ee_link_name_);
  xdot_ = robot_data_->getVelocity(ee_link_name_);
}

std::unordered_map<std::string, double> FR3Controller::compute() 
{
  // One-time init per mode entry (snapshot current measured states)
  if (is_mode_changed_) 
  {
    is_mode_changed_ = false;
    control_start_time_ = sim_time_;

    q_init_ = q_;
    qdot_init_ = qdot_;
    x_init_ = x_;
    xdot_init_ = xdot_;

    // Reset desired trajectories to snapshots
    q_desired_ = q_init_;
    qdot_desired_.setZero(dof_);
    x_desired_ = x_init_;
    xdot_desired_.setZero();
  }

  // --- Mode: Home (joint-space cubic to a predefined posture) ---
  if (control_mode_ == "Home") {
    Eigen::Vector7d q_home;
    q_home << 0.0, 0.0, 0.0, -M_PI/2., 0.0, M_PI/2., M_PI / 4.;

    q_desired_ = robot_controller_->moveJointPositionCubic(q_home,
                                                           Eigen::VectorXd::Zero(dof_),
                                                           q_init_,
                                                           qdot_init_,
                                                           sim_time_,
                                                           control_start_time_,
                                                           3.0);

    qdot_desired_ = robot_controller_->moveJointVelocityCubic(q_home,
                                                              Eigen::VectorXd::Zero(dof_),
                                                              q_init_,
                                                              qdot_init_,
                                                              sim_time_,
                                                              control_start_time_,
                                                              3.0);

    tau_desired_ = robot_controller_->moveJointTorqueStep(q_desired_, qdot_desired_);

    // Alternative: synthesize torque over time with a single API
    // tau_desired = robot_controller->moveJointTorqueCubic(...)
  }
  // --- Mode: QPIK (task-space, QP-based IK with cubic profiling) ---
  else if (control_mode_ == "QPIK") 
  {
    Eigen::Affine3d target_x = x_init_;
    target_x.translation() += Eigen::Vector3d(0.0, 0.1, 0.1); // +10 cm in Y and Z

    qdot_desired_ = robot_controller_->QPIKCubic(target_x, 
                                                 Eigen::VectorXd::Zero(6),
                                                 x_init_, 
                                                 xdot_init_,
                                                 sim_time_, 
                                                 control_start_time_, 
                                                 3.0, 
                                                 ee_link_name_);

    // Simple Euler integrate desired joint positions from qdot_desired
    q_desired_   = q_ + qdot_desired_ * dt_;
    tau_desired_ = robot_controller_->moveJointTorqueStep(q_desired_, qdot_desired_);
  }
  // --- Mode: Gravity Compensation (no tracking) ---
  else if (control_mode_ == "Gravity Compensation") 
  {
    tau_desired_ = robot_data_->getGravity();
  }

  // Format output for simulator actuators: joint name -> tau
  std::unordered_map<std::string, double> ctrl_dict;
  ctrl_dict.reserve(dof_);
  for (size_t i = 0; i < dof_; ++i) 
  {
    ctrl_dict.emplace("fr3_joint" + std::to_string(i+1), tau_desired_(i));
  }
  return ctrl_dict;
}

void FR3Controller::setMode(const std::string& control_mode) 
{
  // Switch control mode and trigger per-mode re-initialization
  is_mode_changed_ = true;
  control_mode_ = control_mode;
  std::cout << "Control Mode Changed: " << control_mode_ << std::endl;
}

void FR3Controller::startKeyListener_() 
{
  tty_ok_ = ::isatty(STDIN_FILENO);
  if (!tty_ok_) 
  {
    std::cout << "[FR3Controller] stdin is not a TTY; keyboard control disabled.\n";
    return;
  }
  setRawMode_();
  stop_key_ = false;
  key_thread_ = std::thread(&FR3Controller::keyLoop_, this);

  std::cout << "[FR3Controller] Keyboard: [1]=Home, [2]=QPIK, [3]=Gravity Compensation\n";
}

void FR3Controller::stopKeyListener_() 
{
  if (!tty_ok_) return;
  stop_key_ = true;
  if (key_thread_.joinable()) key_thread_.join();
  restoreTerm_();
}

void FR3Controller::setRawMode_() 
{
  if (!tty_ok_) return;
  struct termios raw;
  if (tcgetattr(STDIN_FILENO, &orig_term_) == -1) 
  {
    perror("tcgetattr");
    tty_ok_ = false;
    return;
  }
  raw = orig_term_;
  raw.c_lflag &= ~(ICANON | ECHO);
  raw.c_cc[VMIN]  = 0;
  raw.c_cc[VTIME] = 1;
  if (tcsetattr(STDIN_FILENO, TCSANOW, &raw) == -1) 
  {
    perror("tcsetattr");
    tty_ok_ = false;
  }
}

void FR3Controller::restoreTerm_() 
{
  if (!tty_ok_) return;
  if (tcsetattr(STDIN_FILENO, TCSANOW, &orig_term_) == -1) 
  {
    perror("tcsetattr restore");
  }
}

void FR3Controller::keyLoop_() 
{
  while (!stop_key_) 
  {
    char c = 0;
    ssize_t n = ::read(STDIN_FILENO, &c, 1);
    if (n == 1) 
    {
      if (c == '1') 
      {
        setMode("Home");
      } else if (c == '2') 
      {
        setMode("QPIK");
      } else if (c == '3') 
      {
        setMode("Gravity Compensation");
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }
  }
}
